doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,617 @@
+- Whether source addresses will be provided to the opposite BIO when it
+  receives datagrams.
+- Whether destination addresses will be provided to the opposite BIO
+  when it receives datagrams.
+
+The usage is as follows:
+
+- One side of a BIO pair calls `BIO_dgram_set_caps` with zero or
+  more of the following flags to advertise its capabilities:
+  - `BIO_DGRAM_CAP_HANDLES_SRC_ADDR`
+  - `BIO_DGRAM_CAP_HANDLES_DST_ADDR`
+  - `BIO_DGRAM_CAP_PROVIDES_SRC_ADDR`
+  - `BIO_DGRAM_CAP_PROVIDES_DST_ADDR`
+- The other side of the BIO pair calls `BIO_dgram_get_effective_caps`
+  to learn the effective capabilities of the BIO. These are the capabilities set
+  by the opposite BIO.
+- The above process can also be repeated in the opposite direction.
+
+#### Local Address Support
+
+**See also:**
+[BIO_s_dgram_pair(3)](https://www.openssl.org/docs/manmaster/man3/BIO_s_dgram_pair.html)
+
+Support for local addressing (the reception of destination addresses for
+incoming packets, and the specification of source addresses for outgoing
+packets) varies by OS. Thus, it may not be available in all circumstances. A
+feature negotiation mechanism is introduced to facilitate this.
+
+`BIO_dgram_get_local_addr_cap` (`BIO_CTRL_DGRAM_GET_LOCAL_ADDR_CAP`) determines
+if a BIO is potentially capable of supporting local addressing on the current
+platform. If it determines that support is available, local addressing support
+must then be explicitly enabled via `BIO_dgram_set_local_addr_enable`
+(`BIO_CTRL_DGRAM_SET_LOCAL_ADDR_ENABLE`). If local addressing support has not
+been enabled, attempts to use local addressing (for example via `BIO_sendmmsg`
+or `BIO_recvmmsg` with a `BIO_MSG` with a non-NULL `local` field) fails.
+
+An explicit enablement call is required because setting up local addressing
+support requires system calls on most operating systems prior to sending or
+receiving packets and we do not wish to do this automatically inside the
+`BIO_sendmmsg`/`BIO_recvmmsg` fastpaths, particularly since the process of
+enabling support could fail due to lack of OS support, etc.
+
+`BIO_dgram_get_local_addr_enable` (`BIO_CTRL_DGRAM_GET_LOCAL_ADDR_ENABLE`) is
+also available.
+
+It is important to note that `BIO_dgram_get_local_addr_cap` is entirely distinct
+from the application capability negotiation mechanism discussed above. Whereas
+the capability negotiation mechanism discussed above allows *applications* to
+signal what they are capable of handling in their usage of a given BIO,
+`BIO_dgram_local_addr_cap` allows a *BIO implementation* to indicate to the
+users of that BIO whether it is able to support local addressing (where
+enabled).
+
+#### `BIO_s_dgram_pair`
+
+**See also:**
+[BIO_s_dgram_pair(3)](https://www.openssl.org/docs/manmaster/man3/BIO_s_dgram_pair.html)
+
+A new BIO implementation, `BIO_s_dgram_pair`, is provided. This is similar to
+the existing BIO pair but provides datagram semantics. It provides full support
```

Comment:
It's unclear to me why we need something that preserves message boundaries. As far as I know, QUIC does not preserve any boundaries, and just like TCP is a stream.

It's also not really clear what datagram semantics mean.

Reply:
The application using QUIC doesn't care about message boundaries. But the underlying BIO absolutely must preserve message boundaries. QUIC is based on UDP that has that property. Any BIO that we use to transmit QUIC packets must behave the same way.